ci/bootstrap.py

```diff
@@ -0,0 +1,336 @@
+
+import kubernetes_asyncio as kube
+
+from hailtop.utils import check_shell_output
+
+from ci.build import BuildConfiguration, Code
+from ci.github import clone_or_fetch_script
+from ci.utils import generate_token
+
+from batch.driver.k8s_cache import K8sCache
+
+KUBERNETES_SERVER_URL = os.environ['KUBERNETES_SERVER_URL']
+
+
+def populate_secret_host_path(host_path, secret_data):
+    os.makedirs(host_path)
+    if secret_data is not None:
+        for filename, data in secret_data.items():
+            with open(f'{host_path}/{filename}', 'wb') as f:
+                f.write(base64.b64decode(data))
+
+
+class LocalJob:
+    def __init__(self, index, image, command, *,
+                 env=None, mount_docker_socket=False, secrets=None, service_account=None, attributes=None, parents=None,
+                 input_files=None, output_files=None,
+                 **kwargs):
+        self._index = index
+        self._image = image
+        self._command = command
+
+        self._env = env
+        self._mount_docker_socket = mount_docker_socket
+        self._parents = parents
+        self._secrets = secrets
+        self._service_account = service_account
+        self._attributes = attributes
+        self._input_files = input_files
+        self._output_files = output_files
+        self._kwargs = kwargs
+
+        self._done = False
+        self._succeeded = None
+
+
+async def docker_run(*args):
+    script = ' '.join([shq(a) for a in args])
+    outerr = await check_shell_output(script)
+
+    cid = outerr[0].decode('ascii').strip()
+
+    outerr = await check_shell_output(f'docker wait {cid}')
+
+    exit_code = int(outerr[0].decode('ascii').strip())
+    return cid, exit_code == 0
+
+
+class LocalBatchBuilder:
+    def __init__(self, attributes, callback):
+        self._attributes = attributes
+        self._callback = callback
+        self._jobs = []
+
+    @property
+    def attributes(self):
+        return self._attributes
+
+    @property
+    def callback(self):
+        return self._callback
+
+    def create_job(self, image, command, **kwargs):
+        index = len(self._jobs)
+        job = LocalJob(index, image, command, **kwargs)
+        self._jobs.append(job)
+        return job
+
+    async def run(self):
+        cwd = os.getcwd()
+        assert cwd.startswith('/')
+
+        batch_token = self._attributes['token']
+        root = f'{cwd}/_/{batch_token}'
+
+        os.makedirs(f'{root}/shared')
+
+        prefix = f'gs://dummy/build/{batch_token}'
+
+        for j in self._jobs:
+            job_name = j._attributes.get('name')
+
+            print(f'{j._index}: {job_name}: running...')
+
+            if j._parents:
+                for p in j._parents:
+                    assert p._done
+                    if not p._succeeded:
+                        print(f'{j._index}: {job_name}: SKIPPED: parent {p._index} failed')
+                        j._done = True
+                        j._failed = True
```

Comment:
It looks like we still try to copy inputs if this happens. Shouldn't we give up on this job at this point?